src/array_like.rs

```diff
@@ -0,0 +1,81 @@
+//! TODO
+
+use crate::{get_array_module, ndarray::Dimension, Element, PyReadonlyArray};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{types::PyDict, FromPyObject, PyAny, PyResult};
+use std::ops::Deref;
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(PyReadonlyArray<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+impl<'py, T, D> Deref for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    type Target = PyReadonlyArray<'py, T, D>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'py, T, D> FromPyObject<'py> for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+    Vec<T>: FromPyObject<'py>,
+{
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        if let Ok(array) = ob.extract::<PyReadonlyArray<T, D>>() {
+            return Ok(PyArrayLike(array));
+        }
+
+        let py = ob.py();
+
+        // The following does not work because of dimension type conflicts.
+        // Maybe we should implement `into_dimensionality` for PyArray?
+        // Or should we just delete that block, since the `asarray` fallback
+        // already takes care of that case?
+
+        // if matches!(D::NDIM, None | Some(1)) {
+        //     if let Ok(array) = ob.extract::<Vec<T>>() {
+        //         let res = PyArray1::from_vec(py, array).readonly();
+        //         return Ok(PyArrayLike(res));
+        //     }
+        // }
+
+        let numpy_module = get_array_module(py)?;
```

Comment:
I think it would be nice to cache the method reference using [`GILOnceCell`](https://docs.rs/pyo3/latest/pyo3/sync/struct.GILOnceCell.html).

Also, do we need to set the `dtype` keyword argument? I think this would try harder than your previous code would?

Reply:
> I think it would be nice to cache the method reference using [`GILOnceCell`](https://docs.rs/pyo3/latest/pyo3/sync/struct.GILOnceCell.html).

That's nice! I was not aware of that Cell type.

> Also, do we need to set the `dtype` keyword argument? I think this would try harder than your previous code would?

I think it is essential to set the `dtype`. E.g. if the input array is an ndarray of dtype `int32` then `np.asarray(...)` wouldn't do anything, so we wouldn't be able to extract a `PyReadonlyArray<f64, D>` from it.

Reply:
> E.g. if the input array is an ndarray of dtype int32 then np.asarray(...) wouldn't do anything, so we wouldn't be able to extract a PyReadonlyArray<f64, D> from it.

But this is exactly what I meant, your direct-line code using `.extract::<Vec<T>>` will also not convert `i32` to `f64` if `T` is `f64`. Hence I would like to avoid making `PyArrayLike` do two different things, turning non-array storage in array storage _and_ converting element types.

Reply:
> > E.g. if the input array is an ndarray of dtype int32 then np.asarray(...) wouldn't do anything, so we wouldn't be able to extract a PyReadonlyArray<f64, D> from it.
> 
> But this is exactly what I meant, your direct-line code using `.extract::<Vec<T>>` will also not convert `i32` to `f64` if `T` is `f64`. Hence I would like to avoid making `PyArrayLike` do two different things, turning non-array storage in array storage _and_ converting element types.

Actually `.extract::<Vec<f64>>` does convert `i32` to `f64` and everything else which is meaningful (it will not convert `f64` to `i32` e.g.).

For me, the whole point of the `PyArrayLike` type is to obviate the need for calling `np.asarray` on Python side before calling a PyO3 function. So it should be able to match against anything which can be reasonably converted into the target type.

Reply:
> Actually .extract::<Vec<f64>> does convert i32 to f64 and everything else which is meaningful (it will not convert f64 to i32 e.g.).

This particular case is an idiosyncrasy of [`PyFloat_AsDouble`](https://docs.python.org/3/c-api/float.html#c.PyFloat_AsDouble) falling back to `__index__` but `FromPyObject` will generally not do lossy conversions and coerce types, whereas

```python
>>> np.asarray([1.5, 2.5, 3.5], dtype=np.int32)
array([1, 2, 3], dtype=int32)
```

happily throws away the fractional parts. As it is written now, passing `[1.5, 2.5, 3.5]` as `PyArrayLike1<i32>` will fall through the `Vec` extract just to be made work by the falling to `asarray` with `dtype` set.

I am not saying, to coercing the data type has no usage, but the above seems inconsistent to me and if you really need lists of floatings as arrays of integers I would propose separate types for the coercing and non-coercing variants and to leave out the `.extract`-based code path for the coercing one.

Not sure about reasonable names though? Since our MSRV allows const generics by now, we could even using a `const COERCE: bool` generic parameter?

Reply:
It is unfortunate that `np.asarray` does "unsafe" casting (like f64 to i32) - didn't know that. For me it is completely fine if we restrict to "safe" casting. Consider my example function from #382:

```rust
#[pyfunction]
fn my_mean(ar: ArrayLike1<f64>) -> PyResult<f64> {
    let res = ar
        .as_array()
        .mean()
        .ok_or(PyValueError::new_err("Empty array"))?;
    Ok(res)
}
```
To me it is very desirable that this kind of function also works e.g. for `np.array([1,2,3,4], dtype="int")`.

I'd suggest to replace our fallback with this python function:
```python3
np.asarray(input_array).astype(target_type, casting='safe')
```
This should hopefully work as intended.

Reply:
> I'd suggest to replace our fallback with this python function:

This seems reasonable. The implementation needs a bit more effort probably, e.g. I would recommend to use `Python::eval` to bind this expression to a single function like

```rust
py.eval("lambda array, dtype: np.asarray(input_array).astype(target_type, casting='safe')", ...);
```

and caching that using `GILOnceCell` instead of constructing the call chain manually using PyO3 interfaces.

Reply:
There is still one caveat. Consider the following snippet:
```python3
a = np.array([1,2,3,4])
np.asarray(a).astype('int32', casting='safe')
np.asarray(a).astype('uint64', casting='safe')
```
Both conversions will fail because `.asarray(...)` converts integer lists to int64 arrays, and int64 cannot be casted to either int32 or uint64 according to the safe-rule.

I think I could live with that behavior, as I could simply use `PyArrayLike<i64, D>` everywhere int arrays are involved. But it doesn't feel ideal.

What do you think?

Reply:
The casting behaviour seems correct to me as `int64` cannot fit `int32` or `uint64`. `FromPyObject` should similarly not support this conversion.

Personally, I would like to avoid open-coding this ourselves. Hence, if you are unhappy with the safe-casting rules, I propose we go back to a `const COERCE: bool` generic parameter to determine whether to pass `dtype` to `numpy.asarray` or not.

Reply:
One might also argue however that an `i64` value _can_ fit into `u64` or `i32`, so we should try to cast (and fail safely in case of too large values). This also seems to be the philosophy of `.extract::<Vec<T>>`. If we decide against trying to cast, we should also drop our 1d-fallback for consistency reasons.

Please see my most recent commit, where I sketched the "two fallbacks" approach. If you dislike it, it is completely fine for me to hide the second fallback behind a `COERCE` flag (and we should also think about dropping the 1d-case as mentioned). I my opinion however, this is the right way to go.

Reply:
I am not happy with the amount of subtlety we are sinking into this API.

So, let's remove the additional 1d fallback based on `.extract::<Vec<T>>()` for consistency and make this a simple convenience/efficiency wrapper for `numpy.asarray`:

1. Try to downcast to `PyArray` and call `.readonly()` instead of `.extract::<PyReadonlyArray>()` because the error path of `FromPyObject` is necessarily slow.
2. Call `numpy.asarray`, either with or without `dtype` passed depending on a `COERCE` type parameter defaulting to `false`.

(Personally, I don't feel we need to remove the 1d fallback because inconsistency there is harmless but I definitely want to bias this towards simple semantics so I wont argue that one.)

Reply:
>     2. Call `numpy.asarray`, either with or without `dtype` passed depending on a `COERCE` type parameter defaulting to `false`.

Just to be clear: We already noticed that setting `dtype` in `np.asarray` will allow casting floats to ints by cutting of precision. Do you really want to allow that (even behind a flag)? I would rather prevent losing precision, but allow safe casting like `i64` to `PyResult<i32>`. This is what my current implementation does. It might be cluttered, but this is the best solution I came up so far.

Reply:
Yes, `rust-numpy` is a Rust wrapper for NumPy hence I would like to it to wrap the NumPy API, not invent its own semantics.

Special handling of `dtype` appears reasonable to me since this is a type parameter on the Rust side, hence the idea of exposing both semantics via a `COERCE` flag.

So here is the design space around https://github.com/PyO3/rust-numpy/pull/383#discussion_r1257289547 that I am comfortable with:

* Include the `COERCE` flag or don't, but without it `asarray` should be called without `dtype`.
* Have the 1d code path via `.iter()` and `.extract::<T>()`, independently of whether `COERCE` is used or not. I am fine with the small inconsistency that this does more than `asarray` without `dtype` passed.

Reply:
> Yes, `rust-numpy` is a Rust wrapper for NumPy hence I would like to it to wrap the NumPy API, not invent its own semantics.

I understand your point, but I only see disadvantages with this semantics:

As a library creator using `rust-numpy`, I would never use `PyArrayLike<i32, COERCE=false>`, `PyArrayLike<u64, COERCE=false>`, `PyArrayLike<f32, COERCE=false>`, ... since these don't match against the default python list types. This would be totally annoying for my users (mostly for those using Python as a REPL).

On the other hand, using `PyArrayLike<i32, COERCE=true>` might be even worse as it has the potential to distort the input array completely without any warning (note that the (Python) users of my library cannot even see/manipulate the COERCE flag). As a result, my library could produce wrong results.

In the end, I would only ever use `PyArrayLike<i64>` and `PyArrayLike<f64>` - and even in this case I might bother people using arrays of dtype float128.

Reply:
I can only re-iterat: I do _not_ want to maintain rust-numpy-specific semantics for `PyArrayLike`. I am willing to include the 1d code path using `FromPyObject` which gives some leeway via `PyFloat_AsDouble`, but that's it.

> In the end, I would only ever use PyArrayLike<i64> and PyArrayLike<f64>

This sounds reasonable to me if the aim is convenience.

Reply:
> (note that the (Python) users of my library cannot even see/manipulate the COERCE flag)

Having the element type as a generic parameter is a core design principle for the current `rust-numpy` and it definitely does chafe against Python API conventions, but it allows the Rust code to be statically checked and optimized.

In practice, this leads to entry points which dispatch into different instantiations of one or more generic functions which is also how a `COERCE` generic parameter could be reified as a function parameter on the Python side.

Reply:
> Having the element type as a generic parameter is a core design principle for the current `rust-numpy` and it definitely does chafe against Python API conventions, but it allows the Rust code to be statically checked and optimized.

I was not criticizing that design at all. I just wanted to point out that the `COERCE`-flag is opaque to my users, so I would rather not put potentially "dangerous" code behind it.

> I can only re-iterate: I do _not_ want to maintain `rust-numpy`-specific semantics for `PyArrayLike`.

Ok, thanks for clarifying. But please understand that I don't see the point in developing a feature that I would probably never use. I would rather create my personal crate (I am not using internals of `rust-numpy` after all), and maybe publish it some day.

I am still very grateful for your patience and all your help: I certainly learned a lot from our discussions. Thanks a lot!

Reply:
> I was not criticizing that design at all. I just wanted to point out that the COERCE-flag is opaque to my users, so I would rather not put potentially "dangerous" code behind it.

I did not interpret your statements that way. I made that reference to point out that the `T` and `D` in a `#[pyfunction]` taking an `PyArray<T, D>` argument are similarly not visible to Python callers. If you want to provide a Python API which does handle them dynamically, you need multiple instantiations (or the `IxDyn` overhead). `COERCE` is no different in that regard.

> I am still very grateful for your patience and all your help: I certainly learned a lot from our discussions. Thanks a lot!

Fair enough. Do you mind if I try to bring this into a shape I am happy to merge and maintain even if you do not use it yourself?

Reply:
That's perfectly fine for me. Please go ahead as you like.

Reply:
Ok, I pushed what I think I will go with. Still needs docs and examples though.